File: CPP/String/D18_SearchPattern.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<int> computeLPS(string pat)
{
      int m = pat.size();
      vector<int> lps(m, 0);
      int len = 0;

      for (int i = 1; i < m;)
      {
            if (pat[i] == pat[len])
            {
                  lps[i++] = ++len;
            }
            else if (len != 0)
            {
                  len = lps[len - 1];
            }
            else
            {
                  lps[i++] = 0;
            }
      }
      return lps;
}

vector<int> KMPSearch(string txt, string pat)
{
      vector<int> result;
      vector<int> lps = computeLPS(pat);
      int i = 0, j = 0, n = txt.size(), m = pat.size();

      while (i < n)
      {
            if (txt[i] == pat[j])
            {
                  i++;
                  j++;
            }

            if (j == m)
            {
                  result.push_back(i - j);
                  j = lps[j - 1];
            }
            else if (i < n && txt[i] != pat[j])
            {
                  if (j != 0)
                  {
                        j = lps[j - 1];
                  }
                  else
                  {
                        i++;
                  }
            }
      }

      return result;
}

int main()
{
      string txt = "aabaacaadaabaaba";
      string pat = "aaba";
      vector<int> indices = KMPSearch(txt, pat);
      for (int idx : indices)
      {
            cout << idx << " ";
      }
      cout << endl;
      return 0;
}

File: Python/String/D18_SearchPattern.py
def computeLPS(pat):
    lps = [0] * len(pat)
    length = 0
    i = 1

    while i < len(pat):
        if pat[i] == pat[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1
    return lps


def KMPSearch(txt, pat):
    lps = computeLPS(pat)
    result = []
    i = 0  
    j = 0  

    while i < len(txt):
        if txt[i] == pat[j]:
            i += 1
            j += 1

        if j == len(pat):
            result.append(i - j)
            j = lps[j - 1]  
        elif i < len(txt) and txt[i] != pat[j]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1  
    return result


if __name__ == "__main__":
    txt = "aabaacaadaabaaba"
    pat = "aaba"
    print("Match indices:", KMPSearch(txt, pat))  
